### PR TITLE
Add progress bar support to to-do items

### PR DESCRIPTION
**Pull request by Akselmo**
As per discussion in this issue https://github.com/pbek/QOwnNotes/issues/2998

I managed to get it to work, tested it out and so far no regressions! :)

This would help my workflow a lot, so I hope you like it as well.

If I add tag support, I'll do it in another PR, however I find progress bar way more useful.

I was also thinking of adding it as a sortable item to the list of tasks, but this now just adds the basic functionality.

image

## Changes

- Maximizing the slider turns the completion checkbox on when saved.
- On hover, there's tooltip for the progress.
- I added tickmarks to the priority slider so it's easier to distinguish between the two
- Clicking on checkbox sets the progress to 100 or 0
- Depending on progress, the completion date is saved or removed

## Todos
- [ ] If subtask or top-level task with subtasks progress is set 100, the upload fails. Clicking on complete checkbox works however.

### CODE REVIEW

**Comment by Akselmo** [comment]
Added screenshot and more detail

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,7 +1,7 @@
 cmake_minimum_required(VERSION 3.2.2)
 
 project(QOwnNotes)
-
+set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
```
Comment (pbek) [inline CMakeLists.txt:4]: Why was this added?

File: .gitignore [hunk 2]
```diff
@@ -27,3 +27,4 @@ ui_*.h
 /src/QOwnNotes
 secrets/*
 
+.cache/clangd/index
```
Comment (pbek) [inline .gitignore:30]: Is this change needed?

File: CMakeLists.txt (hunk 1 above)
Comment (Akselmo) [reply CMakeLists.txt:4]: Oops I forgot I added that lol, I'll remove it

File: src/dialogs/tododialog.ui [hunk 3]
```diff
@@ -105,16 +105,19 @@
             <property name="orientation">
              <enum>Qt::Horizontal</enum>
             </property>
+            <property name="tickPosition">
```
Comment (pbek) [inline src/dialogs/tododialog.ui:108]: What are those new 3 lines for?

File: .gitignore (hunk 2 above)
Comment (Akselmo) [reply .gitignore:30]: Nope, another mishap by me, will remove

File: CMakeLists.txt (hunk 1 above)
Comment (Akselmo) [reply CMakeLists.txt]: Removed it now

File: .gitignore (hunk 2 above)
Comment (Akselmo) [reply .gitignore]: Fixed

File: src/dialogs/tododialog.ui (hunk 3 above)
Comment (Akselmo) [reply src/dialogs/tododialog.ui:108]: This adds the tickmarks to the priority slider so it is easier to distinguish between priority and progress slider by glance. Check the screenshot in PR summary for what I mean :)

File: src/entities/calendaritem.cpp [hunk 4]
```diff
@@ -486,12 +496,12 @@ bool CalendarItem::store() {
...
             "INSERT INTO calendarItem "
             "(summary, url, description, calendar, uid, related_uid, ics_data, "
-            "etag, last_modified_string, has_dirty_data, completed, "
+            "etag, last_modified_string, has_dirty_data, completed, progress, "
             "alarm_date, priority, created, modified, "
             "completed_date, sort_priority) "
             "VALUES (:summary, :url, :description, :calendar, :uid,"
             " :related_uid, :ics_data, :etag, :last_modified_string, "
-            ":has_dirty_data, :completed, :alarm_date, :priority, "
+            ":has_dirty_data, :completed, progress, :alarm_date, :priority, "
```
Comment (pbek) [inline src/entities/calendaritem.cpp:504]: This looks like a `:` is missing. Are you sure this query ever worked? 😉

Comment (Akselmo) [reply src/entities/calendaritem.cpp]: Oops.. I did not get any warnings though :joy: Added it there, will test again now.

**Comment by pbek** [comment]
Did you test your changes with Nextcloud Tasks and mobile applications (like Tasks and jtx Board on Android)? Because this is where most likely regressions will happen.

File: src/entities/calendaritem.cpp (hunk 4 above)
Comment (Akselmo) [reply src/entities/calendaritem.cpp]: I don't see any noticeable difference but at least it's there

**Comment by Akselmo** [comment]
> Did you test your changes with Nextcloud Tasks and mobile applications (like Tasks and jtx Board on Android)? Because this is where most likely regressions will happen.

Good points!

I tested with Nextcloud Tasks. It followed the task status closely to what I set to Qownnotes and vice versa.

On my phone I have Tasks.org and aCalendar+, I will also test it there!

File: src/services/databaseservice.cpp [hunk 5]
```diff
@@ -876,6 +876,11 @@ bool DatabaseService::setupTables() {
         version = 41;
     }
 
+    if (version < 42) {
+        queryDisk.exec(QStringLiteral("ALTER TABLE calendarItem ADD progress INTEGER DEFAULT 0"));
```
Comment (pbek) [inline src/services/databaseservice.cpp:880]: What happens to already completed tasks with this change? You are setting the progress to "0" for those too!

**Comment by pbek** [comment]
> > Did you test your changes with Nextcloud Tasks and mobile applications (like Tasks and jtx Board on Android)? Because this is where most likely regressions will happen.
> 
> Good points!
> 
> I tested with Nextcloud Tasks. It followed the task status closely to what I set to Qownnotes and vice versa.
> 
> On my phone I have Tasks.org and aCalendar+, I will also test it there!

Every app has their different ways to interpret or write the iCal "file". 🙈 I had to build in several fail-saves until it worked "everywhere".

File: src/services/databaseservice.cpp (hunk 5 above)
Comment (Akselmo) [reply src/services/databaseservice.cpp:880]: For me at least, it did not change the progress to 0 for them. The progress data is already there (completed seemed to be just checking if progress is 0 or 100). 

I have old database without any of the changes and when I use it, it updates accordingly, the completed tasks stay completed and their progress slider is just turned to 100.

Comment (Akselmo) [reply src/services/databaseservice.cpp:880]: Tested it again, the priority slider is just set to max on anything that is set completed and it stays completed as it should.

**Comment by Akselmo** [comment]
> Every app has their different ways to interpret or write the iCal "file". 🙈 I had to build in several fail-saves until it worked "everywhere".

Yeah seems Tasks.org does not have progress, nor aCalendar+ (which uses Tasks.org to handle the tasks anyway so we can concentrate on Tasks.org only.)

However, it does keep the old progress value around when I set the completion status off. Both nextcloud tasks and qownnotes show it as it should.

File: src/services/databaseservice.cpp (hunk 5 above)
Comment (pbek) [reply src/services/databaseservice.cpp:880]: If the user doesn't open the todo dialog after a migration, the progress state of 0 could be synced to the CalDAV server. It would be more save when the migration had two more queries...

1) set the progress to 0 if completed is 0
2) set the progress to 100 if completed is 1

**Comment by pbek** [comment]
In different apps...
How does setting progress on setting a parent task influences a child task?
And the other way around, how does setting progress on setting a child task influences the parent task?

I wonder if they are getting confused if the states of the parent doesn't match the child and vice versa... 🤔

**Comment by Akselmo** [comment]
> In different apps... How does setting progress on setting a parent task influences a child task? And the other way around, how does setting progress on setting a child task influences the parent task?
> 
> I wonder if they are getting confused if the states of the parent doesn't match the child and vice versa... 🤔

Seems that child tasks don't show properly in either end for me. However I have had the same issue with the current version of the client, not the one with progress bar.. Seems something breaks with them?

For example creating todo with subtask on qownnotes doesnt show on nextcloud tasks, but if i do the same on nextcloud tasks, i see both tasks on qownnotes, but they're not connected to each other and shown as separate tasks.

File: src/services/databaseservice.cpp (hunk 5 above)
Comment (Akselmo) [reply src/services/databaseservice.cpp:880]: Good idea. I'll add that... So something like this?

```cpp
    if (version < 42) {
        queryDisk.exec(QStringLiteral("ALTER TABLE calendarItem ADD progress INTEGER DEFAULT 0"));
        // loop over all items in calendarItem table
        // if completed is 0, set progress 0
        // if completed is 1, set progress 100
        version = 42;
    }

```

I am not very familiar with databases so how can I do this? :sweat_smile:

**Comment by pbek** [comment]
> For example creating todo with subtask on qownnotes doesnt show on nextcloud tasks, but if i do the same on nextcloud tasks, i see both tasks on qownnotes, but they're not connected to each other and shown as separate tasks.

I just created a new task with sub-task in QOwnNotes (current release). It shows up for me in Nextcloud Tasks. 🤔

image

File: src/services/databaseservice.cpp (hunk 5 above)
Comment (pbek) [reply src/services/databaseservice.cpp:880]: It's much easier: 😉

```sql
UPDATE calendarItem SET progress = 0 WHERE completed = 0;
UPDATE calendarItem SET progress = 100 WHERE completed = 1;
```

**Comment by Akselmo** [comment]
> I just created a new task with sub-task in QOwnNotes (current release). It shows up for me in Nextcloud Tasks. 🤔

Okay I think something is broken with my caldav then, I have a task that references itself as subtask (i have no idea how that has happened???) so it just confuses all my clients. Fun.

**Comment by pbek** [comment]
> Okay I think something is broken with my caldav then, I have a task that references itself as subtask (i have no idea how that has happened???) so it just confuses all my clients. Fun.

Phew, I wonder how that happened! 😬

File: src/services/databaseservice.cpp (hunk 5 above)
Comment (Akselmo) [reply src/services/databaseservice.cpp:880]: Thanks, I added those there. :)

**Comment by Akselmo** [comment]
> > Okay I think something is broken with my caldav then, I have a task that references itself as subtask (i have no idea how that has happened???) so it just confuses all my clients. Fun.
> 
> Phew, I wonder how that happened! 😬

Lol no idea.. Anyhow, I fixed it by downloading the calendar, removing the offending item and uploading it again... 

But yeah now I can create tasks just fine again. The progresses seem to be independent per task, although if I set the top level task as 100 in nextcloud, the childtasks get completed too. Seems something breaks there when trying to set progress to 100 in the task, maybe it's not marked completed properly. Will have to investigate that part.

**Comment by pbek** [comment]
> Seems something breaks there when trying to set progress to 100 in the task, maybe it's not marked completed properly. Will have to investigate that part.

It's very important that it will also be set as completed!

File: src/entities/calendaritem.cpp [hunk 6]
```diff
@@ -77,15 +79,21 @@ void CalendarItem::setAlarmDate(const QDateTime &dateTime) { this->alarmDate = d
 
 void CalendarItem::setPriority(int value) { this->priority = value; }
 
-void CalendarItem::setCompleted(bool value) { this->completed = value; }
+void CalendarItem::setProgress(int value) { this->progress = value; }
```
Comment (pbek) [inline src/entities/calendaritem.cpp:82]: You need to use `setCompleted(value = 100)` too here! But don't use `setProgress` inside `setCompleted` then, but set the value directly. 😉

Comment (pbek) [reply src/entities/calendaritem.cpp:82]: Hm, but you don't want to override progress too in that case inside setCompleted... you need to check that somewhere else...

Comment (Akselmo) [reply src/entities/calendaritem.cpp:82]: Yeah, I wasn't sure which way to go here. :sweat_smile:

Comment (pbek) [reply src/entities/calendaritem.cpp:82]: Either way, progress and completed need to be "linked" in some way.

**Comment by pbek** [comment]
I tried the PR, two things came up:

1) I think it would be good to see and be able to edit the actual process as a number (like in Nextcloud tasks).
2) When I move the progress slider of a task to 100% and pressed *Save* I got an `QNetworkReply error 299` in the log, the task vanished from the list (I needed to *Clear cache and reload* to get it again) and the progress was not stored.
The Nextcloud log said: `Sabre\DAV\Exception\UnsupportedMediaType: Validation error in iCalendar: The supplied value () is not a correct DATE-TIME`

**Comment by pbek** [comment]
ad 2) I wasn't even able to save any other change in any other task. Without this PR saving the same tasks work.

**Comment by Akselmo** [comment]
Yeah I may give up on the progress slider part for now.. :/ I only really need the tag system personally, so I am concentrating on that now.

Sorry for the extra work I have caused with this. I may pick up on this PR later. (Or invite anyone else to do so!)

**Comment by pbek** [comment]
Then I'll close this for now.
